### PR TITLE
Make CodepointWidthDetector::GetWidth faster

### PR DESCRIPTION
**Pull request by skyline75489**
This is a subset of #3578 which I think is harmless and the first step towards making things right.
References #3546 #3578 

## Detailed Description of the Pull Request / Additional comments

For more robust Unicode support, `CodepointWidthDetector` should provide concrete width information rather than a simple boolean of `IsWide`. Currently only `IsWide` is widely used and optimized using quick lookup table and fallback cache. This PR moves those optimization into `GetWidth`.

## Validation Steps Performed

API remains unchanged. Things are not broken.

### CODE REVIEW

File: src/types/CodepointWidthDetector.cpp [hunk 1]
```diff
@@ -369,74 +387,67 @@ bool CodepointWidthDetector::IsWide(const wchar_t wch) const noexcept
...
-        {
-            return width == CodepointWidth::Wide;
-        }
+// Routine Description:
+// - returns the width type of codepoint by searching the map generated from the unicode spec
+// Arguments:
+// - glyph - the utf16 encoded codepoint to search for
+// Return Value:
+// - the width type of the codepoint
+CodepointWidth CodepointWidthDetector::_lookupGlyphWidth(const std::wstring_view glyph) const
```
Comment (skyline75489) [inline src/types/CodepointWidthDetector.cpp]: The original `GetWidth` is renamed to `_lookupGlyphWidth` and made private.

**Comment by skyline75489** [comment]
And by the way, isn't it just way more natural to implement `IsWide` as `return GetWidth(glyph) == CodepointWidth::Wide`? Code resue, people. Am I right?

**Comment by zadjii-msft** [comment]
(For the record, as it is Thanksgiving over here in the US, there's a good chance that no one reads this PR till next week. We're not just ghosting it 😋)

**Comment by jsoref** [comment]
@miniksa : out of curiosity, what happened to this PR? (I'm mostly looking through PRs to make sure my action isn't making a mess of things)

**Comment by miniksa** [comment]
> @miniksa : out of curiosity, what happened to this PR? (I'm mostly looking through PRs to make sure my action isn't making a mess of things)

@jsoref, what happened = I spaced and now it's April.

**Comment by DHowett-MSFT** [comment]
Pushed a merge to trigger new CI and friends.

**Comment by DHowett-MSFT** [comment]
Thanks so much for rewriting this guy. :smile:

**Comment by DHowett-MSFT** [comment]
Actually, it looks like `GetWidth` still shares the pfnFallback logic with `_lookup..WithCache` .. is there a chance for us to make this even more streamlined?

**Comment by skyline75489** [comment]
@DHowett-MSFT I wish I could. I think the `pfnFallback` logic is delibrate. It reduces the need for the `_lookupGlyphWidth` call, which is expensive, when there is a fallback cache.

Actually the entire `_fallbackCache` should be rewrited to reflect the actual _width_ information instead of booleans. The rewriting does not really fit in this PR. So I intended to leave it as is.
But rewriting `_fallbackCache` should be what I was heading to.

Honestly this PR feels vague to me at first. Took me a while to figure out what I was trying to accomplish.

**Comment by DHowett-MSFT** [comment]
Thanks for paging it back into your memory. That all makes sense, I’ll mark it for merge :)

**Comment by skyline75489** [comment]
@miniksa @zadjii-msft It's all right. Back then when I wrote this PR I was still trying to crack job interviews. And here we are. Life is like a box of chocolates. You never know what you gonna get. Just like you never know the difference between the meaning of "ghosing" in dictionary and in real life. (just kidding, you know how I enjoy working with you guys 😸 )

**Pull request merged**